File: calculate.py
# Time Complexity: O(n)
# Space Complexity: O(n)

class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        stack, cur, op = [], 0, '+'
        for c in s + '+':
            if c == " ":
                continue
            elif c.isdigit():
                cur = (cur * 10) + int(c)
            else:
                if op == '-':
                    stack.append(-cur)
                elif op == '+':
                    stack.append(cur)
                elif op == '*':
                    stack.append(stack.pop() * cur)
                elif op == '/':
                    stack.append(int(stack.pop() / cur))
                cur, op = 0, c
        return sum(stack)

File: calculate.java
//TC - O(n);
//SC - O(k); k is number of digits;
//LC -227

class Solution {
    public int calculate(String s) {
        Stack<Integer> st = new Stack<>();
        int currNum=0;
        char op = '+';
        
        for(int i=0;i<s.length();i++){
            char currChar = s.charAt(i);
            
            if(Character.isDigit(currChar)){
                currNum = currNum*10 + (currChar-'0');
            }
            
            if((!Character.isDigit(currChar) && !Character.isWhitespace(currChar)) || i == s.length()-1){
                if(op=='+'){
                    st.push(currNum);
                }
                else if(op == '-'){
                    st.push(-currNum);
                }
                else if(op == '*'){
                    st.push(st.pop()*currNum);
                }
                else if(op == '/'){
                    st.push(st.pop()/currNum);
                }
                
                op = currChar;
                currNum = 0;
            }
        }
        
        int res=0;
        while(!st.isEmpty()){
            res=res+st.pop();
        }
        
        return res;
    }
}